### PR TITLE
Fix non-graceful shutdown

### PR DESCRIPTION
**Pull request by Tabrizian**

### CODE REVIEW

File: src/resources/startup.py [hunk 1]
```diff
@@ -319,5 +358,11 @@ def sigterm_handler(signum, frame):
 
     server.add_insecure_port(FLAGS.socket)
     server.start()
+    time.sleep(4)
```
Comment (deadeyegoodwin) [inline src/resources/startup.py:361]: Why wait here on the main thread. Can't you just change the location of the sleep in the background thread?

File: src/resources/startup.py [hunk 2]
```diff
@@ -293,10 +297,45 @@ def Execute(self, request, context):
         return execute_response
 
 
+def watch_connections(address, event):
+    number_of_failures = 0
+    timeout_sec = 1
+
+    while True:
```
Comment (deadeyegoodwin) [inline src/resources/startup.py:304]: So this makes sure that a connection is established within a timeout after the server starts? Is there a particular reason it is done this way? Why wait for 3 failures? If I have success, failure, failure, then things are ok? That's doesn't seem right. If the timeout is 12 seconds, why not just wait 12 seconds and then check once?

File: src/resources/startup.py [hunk 3]
```diff
@@ -293,10 +297,45 @@ def Execute(self, request, context):
...
+        if len(sockets.socket_ref) == 1:
+            number_of_failures += 1
+        else:
+            number_of_failures = 0
+
+        # Sleep for 4 seconds before polling next time
+        time.sleep(4)
+
+
 if __name__ == "__main__":
```
Comment (deadeyegoodwin) [inline src/resources/startup.py:334]: I thought there was also a case where if triton just dies then the python process can be left running (zombie). How does this fix that case?

**Review by deadeyegoodwin — CHANGES_REQUESTED** [review]

File: src/resources/startup.py (hunk 2 above)
Comment (Tabrizian) [reply src/resources/startup.py:304]: The goal here is to check that always there is a socket other than this socket connect to the gRPC server. Two things can happen when there is a problem. First, the Python GRPC server may have stopped. Or the Triton might have exited non-gracefully. If the first case happened, the connection to the Python GRPC will fail. I just added this to make sure that there is not a temporary problem with the connection. I can probably reduce the number of failures to 2. But in the example you mentioned, if you have success, failure, failure, failure, then the Python backend will exit.

File: src/resources/startup.py (hunk 3 above)
Comment (Tabrizian) [reply src/resources/startup.py:334]: Mentioned in the comment above how this fixes it. https://github.com/triton-inference-server/python_backend/pull/10/files/bb86d2ba9ddeba7df1a2dc3b461f7b423addd65e#r496131640

File: src/resources/startup.py (hunk 1 above)
Comment (Tabrizian) [reply src/resources/startup.py:361]: Sure, I will update it and remove it from here.

File: src/resources/startup.py (hunk 2 above)
Comment (deadeyegoodwin) [reply src/resources/startup.py:304]: I see... what you have is ok then.

**Review by deadeyegoodwin — APPROVED** [review]

**Pull request merged**